### PR TITLE
[BugFix] Fixing Span operation names generated from RestActions

### PR DESCRIPTION
**Pull request by atharvasharma61**
### Description
This is a bug fix to have RestActions rawPath as span operation names, query params as their tags and avoid using exact URI (#10791)

### Related Issues
N/A

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/telemetry/tracing/AttributeNames.java [hunk 1]
```diff
@@ -40,6 +40,11 @@ private AttributeNames() {
      */
     public static final String HTTP_URI = "http.uri";
 
+    /**
+     * Http Request Query Path.
+     */
+    public static final String HTTP_REQ_QUERY_PARAMS = "http.query_params";
```
Comment (reta) [inline server/src/main/java/org/opensearch/telemetry/tracing/AttributeNames.java:46]: We would need to use Semantic Conventions for HTTP Spans (https://opentelemetry.io/docs/specs/semconv/http/http-spans/):

```suggestion
    public static final String HTTP_REQ_QUERY_PARAMS = "url.query";
```

File: server/src/main/java/org/opensearch/telemetry/tracing/AttributeNames.java [hunk 2]
```diff
@@ -50,6 +55,12 @@ private AttributeNames() {
      */
     public static final String REST_REQ_RAW_PATH = "rest.raw_path";
 
+    /**
+     * Rest Request Query Params.
+     */
+    public static final String REST_REQ_QUERY_PARAMS = "rest.query_params";
```
Comment (reta) [inline server/src/main/java/org/opensearch/telemetry/tracing/AttributeNames.java:61]: We use REST over HTTP, please use `HTTP_REQ_QUERY_PARAMS` instead

File: server/src/main/java/org/opensearch/telemetry/tracing/SpanBuilder.java [hunk 3]
```diff
@@ -114,9 +127,17 @@ private static String createSpanName(RestRequest restRequest) {
 
     private static Attributes buildSpanAttributes(RestRequest restRequest) {
         if (restRequest != null) {
-            return Attributes.create()
+            Attributes attributes = Attributes.create()
                 .addAttribute(AttributeNames.REST_REQ_ID, restRequest.getRequestId())
                 .addAttribute(AttributeNames.REST_REQ_RAW_PATH, restRequest.rawPath());
+
+            try {
```
Comment (reta) [inline server/src/main/java/org/opensearch/telemetry/tracing/SpanBuilder.java:134]: `restRequest.params()` is the query string, no need to convert to URI

File: server/src/main/java/org/opensearch/telemetry/tracing/SpanBuilder.java [hunk 4]
```diff
@@ -75,7 +76,12 @@ public static SpanCreationContext from(String spanName, String nodeId, Replicate
     }
 
     private static String createSpanName(HttpRequest httpRequest) {
-        return httpRequest.method().name() + SEPARATOR + httpRequest.uri();
+        try {
+            URI uri = new URI(httpRequest.uri());
```
Comment (reta) [inline server/src/main/java/org/opensearch/telemetry/tracing/SpanBuilder.java:80]: We'll be creating `URI` at least twice every time (typical usage below):

```
    public static SpanCreationContext from(HttpRequest request) {
        return SpanCreationContext.server().name(createSpanName(request)).attributes(buildSpanAttributes(request));
    }
```

I think we could borrow a very simple `RestRequest#path` implementation to split uri to path + query string without creating tons of URI class instances.

File: server/src/main/java/org/opensearch/telemetry/tracing/AttributeNames.java [hunk 5]
```diff
@@ -40,6 +40,11 @@ private AttributeNames() {
      */
     public static final String HTTP_URI = "http.uri";
 
+    /**
+     * Http Request Query Path.
```
Comment (reta) [inline server/src/main/java/org/opensearch/telemetry/tracing/AttributeNames.java:44]:
```suggestion
     * Http Request Query Parameters.
```

File: server/src/main/java/org/opensearch/telemetry/tracing/SpanBuilder.java [hunk 6]
```diff
@@ -11,13 +11,15 @@
...
 import org.opensearch.common.annotation.InternalApi;
+import org.opensearch.common.collect.Tuple;
 import org.opensearch.core.common.Strings;
 import org.opensearch.http.HttpRequest;
 import org.opensearch.rest.RestRequest;
 import org.opensearch.telemetry.tracing.attributes.Attributes;
 import org.opensearch.transport.TcpChannel;
 import org.opensearch.transport.Transport;
 
+import java.net.URI;
```
Comment (reta) [inline server/src/main/java/org/opensearch/telemetry/tracing/SpanBuilder.java:22]: This is not needed, right?

File: server/src/main/java/org/opensearch/telemetry/tracing/SpanBuilder.java [hunk 7]
```diff
@@ -84,9 +88,24 @@ private static Attributes buildSpanAttributes(HttpRequest httpRequest) {
             .addAttribute(AttributeNames.HTTP_METHOD, httpRequest.method().name())
             .addAttribute(AttributeNames.HTTP_PROTOCOL_VERSION, httpRequest.protocolVersion().name());
         populateHeader(httpRequest, attributes);
+
+        Tuple<String,String> uriParts = splitUri(httpRequest.uri());
+        String query = uriParts.v2();
+        attributes.addAttribute(AttributeNames.HTTP_REQ_QUERY_PARAMS, query);
```
Comment (reta) [inline server/src/main/java/org/opensearch/telemetry/tracing/SpanBuilder.java:94]:
```suggestion
        if (query.isBlank() == false) {
            attributes.addAttribute(AttributeNames.HTTP_REQ_QUERY_PARAMS, query);
        }
```

File: server/src/main/java/org/opensearch/telemetry/tracing/SpanBuilder.java [hunk 8]
```diff
@@ -114,9 +132,15 @@ private static String createSpanName(RestRequest restRequest) {
...
     private static Attributes buildSpanAttributes(RestRequest restRequest) {
         if (restRequest != null) {
-            return Attributes.create()
+            Attributes attributes = Attributes.create()
                 .addAttribute(AttributeNames.REST_REQ_ID, restRequest.getRequestId())
                 .addAttribute(AttributeNames.REST_REQ_RAW_PATH, restRequest.rawPath());
+
+            Tuple<String,String> uriParts = splitUri(restRequest.uri());
+            String query = uriParts.v2();
+            attributes.addAttribute(AttributeNames.HTTP_REQ_QUERY_PARAMS, query);
```
Comment (reta) [inline server/src/main/java/org/opensearch/telemetry/tracing/SpanBuilder.java:141]:
```suggestion
if (query.isBlank() == false) {
            attributes.addAttribute(AttributeNames.HTTP_REQ_QUERY_PARAMS, query);
}
```

**Comment by reta** [comment]
Thanks @atharvasharma61 , please add some tests, thank you

**Comment by deshsidd** [comment]
LGTM. Please add tests as @reta has suggested and fix the build failures.

**Review by deshsidd — COMMENTED** [review]
LGTM. Please add tests as @reta has suggested and fix the build failures.

File: server/src/test/java/org/opensearch/telemetry/tracing/SpanBuilderTests.java [hunk 9]
```diff
@@ -36,20 +36,22 @@ public void testHttpRequestContext() {
         HttpRequest httpRequest = createHttpRequest();
         SpanCreationContext context = SpanBuilder.from(httpRequest);
         Attributes attributes = context.getAttributes();
-        assertEquals("GET /_test", context.getSpanName());
+        assertEquals("GET /_test/resource", context.getSpanName());
```
Comment (reta) [inline server/src/test/java/org/opensearch/telemetry/tracing/SpanBuilderTests.java:39]: The way test is modified right now - it only checks URIs with query strings, I would suggest to convert it to parameterized test case and exercise different combinations

**Comment by Gaganjuneja** [comment]
I think in the main issue we talked about templated URI. 

```
For example:

curl http://localhost:9200/*/_search 
curl http://localhost:9200/*/_search?track_total_hits=false
curl http://localhost:9200/test51/_search?track_total_hits=false

This calls will generate 3 unique operation names but is has to be only one: GET /{indices}/_search with other details of the invocation becoming the tags.
```
Here, should we keep the name just like `GET /{indices}/_search` and put the indices as tags for standardizing the span names or having the actual indices as part of the SpanName is fine?

E.g
request - curl http://localhost:9200/test51/_search?track_total_hits=false
Option1 - Span Name : GET /{indices}/_search , tags: indices=test51, params="track_total_hits=false"
Option2 - Span Name: GET /test51/_search , tags: params="track_total_hits=false"

**Comment by reta** [comment]
> Here, should we keep the name just like `GET /{indices}/_search` and put the indices as tags for standardizing the span names or having the actual indices as part of the SpanName is fine?

The span name should be `GET /{indices}/_search`, the full URI (part of trace attributes) would provide the full context. But I think this pull request just tried to separate URI from query string (it is not addressing the issue at large but clears URI a bit)

**Comment by Gaganjuneja** [comment]
> > Here, should we keep the name just like `GET /{indices}/_search` and put the indices as tags for standardizing the span names or having the actual indices as part of the SpanName is fine?
> 
> 
> 
> The span name should be `GET /{indices}/_search`, the full URI (part of trace attributes) would provide the full context. But I think this pull request just tried to separate URI from query string (it is not addressing the issue at large but clears URI a bit)

Yes, we can take that out in the another PR but just wanted to confirm the understanding. Thanks @reta.

**Comment by atharvasharma61** [comment]
Will take that up in another PR as discussed offline with @Gaganjuneja. Thanks!

File: server/src/test/java/org/opensearch/telemetry/tracing/SpanBuilderTests.java [hunk 10]
```diff
@@ -32,24 +32,65 @@
 
 public class SpanBuilderTests extends OpenSearchTestCase {
```
Comment (reta) [inline server/src/test/java/org/opensearch/telemetry/tracing/SpanBuilderTests.java:34]: Please use parameterized tests (plenty of examples in the codebase:  

```
    @ParametersFactory
    public static Collection<Object[]> parameters() {
        return Arrays.asList(
            new Object[] { "/_test/resource?name=John&age=25", "GET /_test/resource", "name=John&age=25" },
            new Object[] { "/_test/", "GET /_test/", "" }
        );
    }
```

**Comment by atharvasharma61** [comment]
Screenshot 2024-02-05 at 8 05 00 PM

File: server/src/test/java/org/opensearch/telemetry/tracing/SpanBuilderTests.java [hunk 11]
```diff
@@ -8,6 +8,8 @@
 
 package org.opensearch.telemetry.tracing;
 
+import com.carrotsearch.randomizedtesting.annotations.ParametersFactory;
+import org.junit.runners.Parameterized;
```
Comment (reta) [inline server/src/test/java/org/opensearch/telemetry/tracing/SpanBuilderTests.java:12]: This is not needed

File: server/src/test/java/org/opensearch/telemetry/tracing/SpanBuilderTests.java [hunk 12]
```diff
@@ -27,29 +29,67 @@
 
 import java.io.IOException;
 import java.util.Arrays;
+import java.util.Collection;
 import java.util.List;
 import java.util.Map;
 
 public class SpanBuilderTests extends OpenSearchTestCase {
 
+    @Parameterized.Parameter(0)
```
Comment (reta) [inline server/src/test/java/org/opensearch/telemetry/tracing/SpanBuilderTests.java:38]: This is not needed - you use constructor for injection

**Review by reta — APPROVED** [review]

**Comment by reta** [comment]
@Gaganjuneja looks good to you? thank you

**Review by Gaganjuneja — APPROVED** [review]
Looks good to me as well. Thanks!

**Comment by Gaganjuneja** [comment]
@atharvasharma61, Can you please update the CHANGE.LOG and there are some :server:spotlessJavaCheck issues. Please address.

**Review by reta — APPROVED** [review]

**Comment by Gaganjuneja** [comment]
Flaky test : org.opensearch.backwards.MixedClusterClientYamlTestSuiteIT

**Comment by reta** [comment]
> Flaky test : org.opensearch.backwards.MixedClusterClientYamlTestSuiteIT

I suspect the 2.x branch in not in a good shape :(

**Comment by Gaganjuneja** [comment]
> > Flaky test : org.opensearch.backwards.MixedClusterClientYamlTestSuiteIT
> 
> I suspect the 2.x branch in not in a good shape :(

Yes, I tried rerunning this more than 10 times and still failing.

**Comment by Gaganjuneja** [comment]
@sachinpkale @gbbafna Could you please merge this. We did a rebase to pass the test cases. Thanks!

**Review by sachinpkale — APPROVED** [review]

**Pull request merged**